crates/bevy_render/src/surface/viewport.rs

```diff
@@ -0,0 +1,223 @@
+                top: SideLocation::Relative(0.0),
+                bottom: SideLocation::Relative(1.0),
+            },
+            scale_factor: 1.0,
+            depth_range: 0.0..=1.0,
+        }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Reflect)]
+#[reflect_value(PartialEq)]
+pub enum SideLocation {
+    Relative(f32),
+    Absolute(f32),
+}
+
+impl Default for SideLocation {
+    fn default() -> Self {
+        Self::Relative(0.0)
+    }
+}
+
+impl Add<f32> for SideLocation {
+    type Output = SideLocation;
+
+    fn add(self, rhs: f32) -> Self::Output {
+        match self {
+            SideLocation::Relative(value) => SideLocation::Relative(value + rhs),
+            SideLocation::Absolute(value) => SideLocation::Absolute(value + rhs),
+        }
+    }
+}
+
+impl Sub<f32> for SideLocation {
+    type Output = SideLocation;
+
+    fn sub(self, rhs: f32) -> Self::Output {
+        match self {
+            SideLocation::Relative(value) => SideLocation::Relative(value - rhs),
+            SideLocation::Absolute(value) => SideLocation::Absolute(value - rhs),
+        }
+    }
+}
+
+impl AddAssign<f32> for SideLocation {
+    fn add_assign(&mut self, rhs: f32) {
+        match self {
+            SideLocation::Relative(value) => *value += rhs,
+            SideLocation::Absolute(value) => *value += rhs,
+        }
+    }
+}
+
+impl SubAssign<f32> for SideLocation {
+    fn sub_assign(&mut self, rhs: f32) {
+        match self {
+            SideLocation::Relative(value) => *value -= rhs,
+            SideLocation::Absolute(value) => *value -= rhs,
+        }
+    }
+}
+
+pub fn viewport_system(
+    mut window_resized_events: EventReader<WindowResized>,
+    mut window_scale_change_events: EventReader<WindowScaleFactorChanged>,
+    windows: Res<Windows>,
+    mut queries: QuerySet<(Query<&Viewport, Changed<Viewport>>, Query<&mut Viewport>)>,
+) {
+    let mut changed_window_ids: HashSet<WindowId> = HashSet::default();
+    for event in window_resized_events.iter() {
+        changed_window_ids.insert(event.id);
+    }
+    for event in window_scale_change_events.iter() {
+        changed_window_ids.insert(event.id);
+    }
+    for viewport in queries.q0().iter() {
+        if let Some(id) = viewport.surface.get_window() {
+            changed_window_ids.insert(id);
+        }
+    }
+
+    // update the surfaces
+    for mut viewport in queries.q1_mut().iter_mut() {
+        match viewport.surface {
+            SurfaceId::Window(id) => {
+                if changed_window_ids.contains(&id) {
+                    let window = windows
+                        .get(id)
+                        .expect("Viewport surface refers to non-existent window");
+                    viewport.update_rectangle(vec2(window.width(), window.height()));
+                    viewport.scale_factor = window.scale_factor();
+                }
+            }
+            SurfaceId::Texture(_id) => {
+                // TODO: not implemented yet
```

Comment:
Does this need to be fixed before merging?

Reply:
Right now there's no offscreen equivalent of `WindowTextureNode`, so really no way of implementing this. That's something to come in a future PR.

That said, I'm really torn on whether to leave it as a `TextureId` for now (which will probably be wrong) or changing it to a `Texture<Handle>` right away, risking being wrong again.